frontend/app/psi.ts

```diff
@@ -0,0 +1,104 @@
+import { base64 } from "rfc4648";
+
+const sodium = require("libsodium-wrappers-sumo");
+
+type ServerResponse = {
+    userPassword: string;
+    breachedPasswords: string[];
+};
+
+export function hashToPoint(input: string): Uint8Array {
+    const hash = sodium.crypto_generichash(
+        sodium.crypto_core_ristretto255_HASHBYTES,
+        sodium.from_string(input)
+    );
+    return sodium.crypto_core_ristretto255_from_hash(hash);
+}
+
+/**
+ *
+ * @param input the string to be encrypted
+ * @returns input with a secret key applied and the key's inverse
+ */
+export function applySeed(input: string): [Uint8Array, Uint8Array] {
+    // generate random seed
+    const seed = sodium.crypto_core_ristretto255_scalar_random();
+    // get seed inverse
+    const seedInverse =
+        sodium.crypto_core_ristretto255_scalar_invert(seed);
+    const point = hashToPoint(input);
+    // apply seed
+    const seededPassword = sodium.crypto_scalarmult_ristretto255(
+        seed,
+        point
+    );
+    console.log("seeded password: ", new TextDecoder().decode(seededPassword.buffer));
+    return [seededPassword, seedInverse];
+}
+
+function computeIntersection(
+    data: ServerResponse,
+    aInverse: Uint8Array
+): boolean {
+    const password = data.userPassword;
+    const serverSet = data.breachedPasswords;
+
+    const options = new Set(
+        serverSet.map(function (element) {
+            return base64.parse(element).join("");
+        })
+    );
+
+    // Client phase 2 - applies inverse seed A to (user password)^ab
+    // so now ((user password)^ab)^-a = (user password)^b
+    const clientPasswordB = sodium.crypto_scalarmult_ristretto255(
+        aInverse,
+        base64.parse(password)
+    );
+    // End of Client phase 2.
+
+    if (options.has(clientPasswordB.join(""))) {
+        console.log(clientPasswordB.join(""));
+        return true;
+    }
+
+    return false;
```

Comment:
**thought**: Make sure this section makes sense. I decode the user password and breached passwords.

Reply:
makes sense to me 👍